### PR TITLE
pg_hba_rule does not properly verify address parameter

### LINKED ISSUE #1373: pg_hab_rule allows any kind of data for address parameter
## Describe the Bug
According to PostgreSQL documentation there are only specific data possible for address:
https://www.postgresql.org/docs/current/auth-pg-hba-conf.html

    IPV4 CIDR
    IPV6 CIDR
    FQDN
    the strings 'samenet' or 'samehost'
    a domain
    a domain with a starting dot

## Expected Behavior
When providing an illegal parameter, I want the catalog compilation to fail

## Steps to Reproduce
Add a parameter like '/12' without IP.
Puppet will add the data to pg_hba_rules file and postgresql service will fail to start.

## Environment
Any version, all platforms.

### PR DESCRIPTION
**Pull request by tuxmea**
According to PostgreSQL documentation there are only specific data possible for address:
https://www.postgresql.org/docs/current/auth-pg-hba-conf.html

- IPV4 CIDR
- IPV6 CIDR
- FQDN
- the strings 'samenet' or 'samehost'
- a domain
- a domain with a starting dot

fixes #1373

### CODE REVIEW

**Review by smortex — COMMENTED** [review]
According to the link in the docstring:

> You can also write **`all` to match any IP address**, `samehost` to match any of the server's own IP addresses, or `samenet` to match any address in any subnet that the server is directly connected to.

File: manifests/server/pg_hba_rule.pp [hunk 1]
```diff
@@ -16,7 +16,16 @@
   String $database,
   String $user,
   String $auth_method,
-  Optional[String] $address       = undef,
+  Optional[
+    Variant[
+      Stdlib::IP::Address::V4::CIDR,
+      Stdlib::IP::Address::V6::CIDR,
+      Stdlib::Fqdn,
+      Enum['samehost', 'samenet'],
```
[the remark this answers is not in the document]
Comment (smortex) [reply manifests/server/pg_hba_rule.pp:24]: That being said, `Stdlib::Fqdn` already match these 3 strings :thinking:

[the remark this answers is not in the document]
Comment (ekohl) [reply manifests/server/pg_hba_rule.pp:24]: I don't mind explicitly calling out these values as special.

**Comment by tuxmea** [comment]
@GSPatton I have issues identifying the error:
Both tests show the following:
```
Failures:

  1) postgresql::server::default_privileges valid object_type schemas on postgres < 10.0 
     Got 0 failures and 2 other errors:

     1.1) Failure/Error: result_facts.merge!(munge_facts(facts)) if self.respond_to?(:facts)

          TypeError:
            no implicit conversion of nil into Hash
          # ./vendor/bundle/ruby/2.7.0/bin/rspec:23:in `load'
          # ./vendor/bundle/ruby/2.7.0/bin/rspec:23:in `<top (required)>'
          # /opt/hostedtoolcache/Ruby/2.7.6/x64/bin/bundle:23:in `load'
          # /opt/hostedtoolcache/Ruby/2.7.6/x64/bin/bundle:23:in `<main>'

     1.2) Failure/Error: @logs.clear

          NoMethodError:
            undefined method `clear' for nil:NilClass
          # ./vendor/bundle/ruby/2.7.0/bin/rspec:23:in `load'
          # ./vendor/bundle/ruby/2.7.0/bin/rspec:23:in `<top (required)>'
          # /opt/hostedtoolcache/Ruby/2.7.6/x64/bin/bundle:23:in `load'
          # /opt/hostedtoolcache/Ruby/2.7.6/x64/bin/bundle:23:in `<main>'

Finished in 7 minutes 24 seconds (files took 1.62 seconds to load)
228 examples, 1 failure

Failed examples:

rspec ./spec/defines/server/default_privileges_spec.rb:132 # postgresql::server::default_privileges valid object_type schemas on postgres < 10.0 
```

**Review by smortex — COMMENTED** [review]
The last PR merged in the main branch had similar CI failures, so these failures seems unrelated to this change.

The change itself looks good to me :+1:

**Review by smortex — APPROVED** [review]

**Comment by tuxmea** [comment]
@david22swan Can you please explain why this is backwards-incompatible?
When specifying an invalid entry, the postgresql service will not start.

**Comment by david22swan** [comment]
Anything that alters what a parameter accepts in such a way as to limit it is technically backwards incompatible.
Since you are limiting it to the correct input only and anything that doesn't match it would have caused errors anyway it is debatable you can probably get it change to bugfix.
Just of the opinion that it needs discussed over first.

File: manifests/server/pg_hba_rule.pp [hunk 2]
```diff
@@ -16,7 +16,16 @@
...
   String $auth_method,
-  Optional[String] $address       = undef,
+  Optional[
+    Variant[
+      Stdlib::IP::Address::V4::CIDR,
+      Stdlib::IP::Address::V6::CIDR,
+      Stdlib::Fqdn,
+      Enum['all', 'samehost', 'samenet'],
+      # RegExp for a DNS domain - also starting with a single dot
+      Pattern[/^\.(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z0-9][a-zA-Z0-9-]{0,61}[a-zA-Z0-9]$/],
```
Comment (david22swan) [inline manifests/server/pg_hba_rule.pp:26]: Any chance that you could convert this to a local type, along the lines of https://github.com/puppetlabs/puppetlabs-apache/blob/main/types/modproxyprotocol.pp. the pattern at least.
As it is, this is kinda a lot

**Comment by ekohl** [comment]
> Anything that alters what a parameter accepts in such a way as to limit it is technically backwards incompatible.

I think I agree with @tuxmea here. Where previously it would be invalid and fail at runtime (starting the service would fail), now it will fail at catalog compilation. I think it could be an enhancement instead of backwards incompatible change.

File: manifests/server/pg_hba_rule.pp (hunk 2 above)
Comment (tuxmea) [reply manifests/server/pg_hba_rule.pp]: Done.

**Comment by bastelfreak** [comment]
I agree with @ekohl and @tuxmea here. I think this is a good enhancement but not a breaking change.

**Comment by tuxmea** [comment]
@david22swan lint fixes have been added.

**Review by david22swan — APPROVED** [review]
LGTM

Gonna do a squash merge since there's a dozen commits, but otherwise this is all good. Thanks for getting back on it so fast.

**Pull request merged**